### PR TITLE
session: del NO_BACKSLASH_ESCAPES sql mode for internal sql

### LINKED ISSUE #43392: Internal SQL give parser error for binary literal when analyze table
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. `git clone https://github.com/pingcap/tidb`
2. `cd tidb && vim session/sessiontest/session_test.go`
3. Append the following code(need to set sql mode **NO_BACKSLASH_ESCAPES**).
```
func TestRandomBinary(t *testing.T) {
    store := testkit.CreateMockStoreWithSchemaLease(t, 1*time.Second)

    tk := testkit.NewTestKit(t, store)
    tk.MustExec("use test")

    ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnStats)
    allBytes := [][]byte{
        {4, 0, 0, 0, 0, 0, 0, 4, '2'},
        {4, 0, 0, 0, 0, 0, 0, 4, '.'},
        {4, 0, 0, 0, 0, 0, 0, 4, '*'},
        {4, 0, 0, 0, 0, 0, 0, 4, '('},
        {4, 0, 0, 0, 0, 0, 0, 4, '\''},
        {4, 0, 0, 0, 0, 0, 0, 4, '!'},
        {4, 0, 0, 0, 0, 0, 0, 4, 29},
        {4, 0, 0, 0, 0, 0, 0, 4, 28},
        {4, 0, 0, 0, 0, 0, 0, 4, 23},
        {4, 0, 0, 0, 0, 0, 0, 4, 16},
    }
    sql := "insert into mysql.stats_top_n (table_id, is_index, hist_id, value, count) values "
    var val string
    for i, bytes := range allBytes {
        if i == 0 {
            val += sqlexec.MustEscapeSQL("(874, 0, 1, %?, 3)", bytes)
        } else {
            val += sqlexec.MustEscapeSQL(",(874, 0, 1, %?, 3)", bytes)
        }
    }
    sql += val
    fmt.Println("gjt debug fmt", sql)
	tk.MustExec("set sql_mode = 'NO_BACKSLASH_ESCAPES';");
    _, err := tk.Session().ExecuteInternal(ctx, sql)
    require.NoError(t, err)
}
```
4. `go test . -run TestRandomBinary --tags=intest -v`

### 2. What did you expect to see? (Required)
run ok
### 3. What did you see instead (Required)
```
[2023/04/26 15:17:18.406 +08:00] [WARN] [session.go:1788] ["parse SQL failed"] [error="line 1 column 307 near \"\u0004\\0\\0\\0\\0\\0\\0\u0004!', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\u001d', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\u001c', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\u0017', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\u0010', 3)\" "] [errorVerbose="line 1 column 307 near \"\u0004\\0\\0\\0\\0\\0\\0\u0004!', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\u001d', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\u001c', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\u0017', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\u0010', 3)\" \ngithub.com/pingcap/errors.AddStack\n\t/home/guojiangtao/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20220729040631-518f63d66278/errors.go:174\ngithub.com/pingcap/errors.Trace\n\t/home/guojiangtao/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20220729040631-518f63d66278/juju_adaptor.go:15\ngithub.com/pingcap/tidb/parser.(*Parser).ParseSQL\n\t/home/guojiangtao/work/tidb/parser/yy_parser.go:170\ngithub.com/pingcap/tidb/session.(*session).ParseSQL\n\t/home/guojiangtao/work/tidb/session/session.go:1565\ngithub.com/pingcap/tidb/session.(*session).ParseWithParams\n\t/home/guojiangtao/work/tidb/session/session.go:1775\ngithub.com/pingcap/tidb/session.(*session).ExecuteInternal\n\t/home/guojiangtao/work/tidb/session/session.go:1677\ngithub.com/pingcap/tidb/session/session_test_test.TestRandomBinary\n\t/home/guojiangtao/work/tidb/session/session_test/session_test.go:4137\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:1576\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1598"] [SQL="insert into mysql.stats_top_n (table_id, is_index, hist_id, value, count) values (874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u00042', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004.', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004*', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004(', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\\'', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004!', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\u001d', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\u001c', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\u0017', 3),(874, 0, 1, _binary'\u0004\\0\\0\\0\\0\\0\\0\u0004\u0010', 3)"]
    session_test.go:4138:
        	Error Trace:	/home/guojiangtao/work/tidb/session/session_test/session_test.go:4138
        	Error:      	Received unexpected error:
        	            	[parser:1064]You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 307 near "\0\0\0\0\0\0!', 3),(874, 0, 1, _binary'\0\0\0\0\0\0', 3),(874, 0, 1, _binary'\0\0\0\0\0\0', 3),(874, 0, 1, _binary'\0\0\0\0\0\0', 3),(874, 0, 1, _binary'\0\0\0\0\0\0', 3)"
        	            	github.com/pingcap/errors.AddStack
        	            		/home/guojiangtao/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20220729040631-518f63d66278/errors.go:174
        	            	github.com/pingcap/errors.(*Error).GenWithStackByArgs
        	            		/home/guojiangtao/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20220729040631-518f63d66278/normalize.go:164
        	            	github.com/pingcap/tidb/util.SyntaxError
        	            		/home/guojiangtao/work/tidb/util/misc.go:159
        	            	github.com/pingcap/tidb/session.(*session).ParseWithParams
        	            		/home/guojiangtao/work/tidb/session/session.go:1790
        	            	github.com/pingcap/tidb/session.(*session).ExecuteInternal
        	            		/home/guojiangtao/work/tidb/session/session.go:1677
        	            	github.com/pingcap/tidb/session/session_test_test.TestRandomBinary
        	            		/home/guojiangtao/work/tidb/session/session_test/session_test.go:4137
        	            	testing.tRunner
        	            		/usr/local/go/src/testing/testing.go:1576
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1598
        	Test:       	TestRandomBinary
```
### 4. What is your TiDB version? (Required)
master cea3c86dfad069bd7b03a4d1164700a7d4b34030

### PR DESCRIPTION
**Pull request by guo-shaoge**
### What problem does this PR solve?

Issue Number: close #43392

Problem Summary: As issue said, `NO_BACKSLASH_ESCAPES` may cause internal sql parse failed.

### What is changed and how it works?
Delete `NO_BACKSLASH_ESCAPES` for internal sql mode.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: session/session.go [hunk 1]
```diff
@@ -1624,8 +1624,11 @@ func (s *session) ClearDiskFullOpt() {
 func (s *session) ExecuteInternal(ctx context.Context, sql string, args ...interface{}) (rs sqlexec.RecordSet, err error) {
 	origin := s.sessionVars.InRestrictedSQL
 	s.sessionVars.InRestrictedSQL = true
+	oriSQLMode := s.sessionVars.SQLMode
+	s.sessionVars.SQLMode = mysql.DelSQLMode(oriSQLMode, mysql.ModeNoBackslashEscapes)
```
Comment (YangKeao) [inline session/session.go:1628]: Why not just set it to default sql mode?

**Review by YangKeao — APPROVED** [review]
LGTM

File: session/session.go (hunk 1 above)
Comment (YangKeao) [reply session/session.go:1628]: Got it. Some function like `checkColumnDefaultValue` depends on `sql_mode`, which will be called by `ExecuteInternal` during bootstrap. If we want to use the default sql_mode, we'll need to review all bootstrap sqls and check whether they can work under the default sql_mode? So, I think it's fine to use `DelSQLMode` :beers: . Good job!

**Review by xhebox — APPROVED** [review]

**Pull request merged**